xen/arch/arm/cpufreq/arm-smc-mailbox.c

```diff
@@ -0,0 +1,234 @@
+/*
+ *  Copyright (C) 2016,2017 ARM Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * This device provides a mechanism for emulating a mailbox by using
+ * smc calls, allowing a "mailbox" consumer to sit in firmware running
+ * on the same core.
+ *
+ * Based on patch series which hasn't reach upstream yet:
+ * => https://lkml.org/lkml/2017/7/23/129
+ *    [PATCH v2 0/3] mailbox: arm: introduce smc triggered mailbox
+ *
+ * Xen modification:
+ * Oleksandr Tyshchenko <Oleksandr_Tyshchenko@epam.com>
+ * Copyright (C) 2017 EPAM Systems Inc.
+ */
+
+#if 0
+#include <linux/device.h>
+#include <linux/kernel.h>
+#include <linux/mailbox_controller.h>
+#include <linux/module.h>
+#include <linux/platform_device.h>
+#include <linux/arm-smccc.h>
+#endif
+
+#include <xen/types.h>
+#include <xen/xmalloc.h>
+#include <xen/err.h>
+
+#include "mailbox_controller.h"
+#include "wrappers.h"
+
+/*
+ * TODO:
+ * 1. Add releasing resources since devm.
+ */
+
+struct arm_smccc_res {
+	unsigned long a0;
+	unsigned long a1;
+	unsigned long a2;
+	unsigned long a3;
+};
+
+/* This is just to align interfaces. */
+static inline void arm_smccc_smc(unsigned long a0, unsigned long a1,
+		unsigned long a2, unsigned long a3, unsigned long a4,
+		unsigned long a5, unsigned long a6, unsigned long a7,
+		struct arm_smccc_res *res)
+{
+	register_t ret[4] = { 0 };
+
+	call_smccc_smc(a0, a1, a2, a3, a4, a5, a6, a7, ret);
+
+	res->a0 = ret[0];
+	res->a1 = ret[1];
+	res->a2 = ret[2];
+	res->a3 = ret[3];
+}
+
+/*
+ * Ugly hack to minimize changes in direct ported code. But, this won't
+ * introduce a bug since we are here only when "use_hvc" flag wasn't set
+ * (smc is only allowed method).
+ */
+#define arm_smccc_hvc arm_smccc_smc
+
+#define ARM_SMC_MBOX_USE_HVC	BIT(0)
+
+struct arm_smc_chan_data {
+	u32 function_id;
+	u32 flags;
+};
+
+static int arm_smc_send_data(struct mbox_chan *link, void *data)
+{
+	struct arm_smc_chan_data *chan_data = link->con_priv;
+	u32 function_id = chan_data->function_id;
+	struct arm_smccc_res res;
+	u32 msg = *(u32 *)data;
+
+	if (chan_data->flags & ARM_SMC_MBOX_USE_HVC)
+		arm_smccc_hvc(function_id, msg, 0, 0, 0, 0, 0, 0, &res);
```

Comment:
As you said ARM_SMC_MBOX_USE_HVC is not expected here. So drop the code, or BUG() here. The "ugly hack" above is really ugly. A cleaner solution would change less lines ;)

Reply:
How does it look like?

static inline void arm_smccc_hvc(unsigned long a0, unsigned long a1,
		unsigned long a2, unsigned long a3, unsigned long a4,
		unsigned long a5, unsigned long a6, unsigned long a7,
		struct arm_smccc_res *res)
{
	/*
	 * We should never get here since the "use_hvc" flag is always false
	 * (smc is only allowed method).
	*/
	BUG();
}